src/cargo/ops/cargo_compile.rs

```diff
@@ -1526,19 +1527,41 @@ fn find_named_targets<'a>(
     };
 
     if proposals.is_empty() {
-        let targets = packages.iter().flat_map(|pkg| {
-            pkg.targets()
-                .iter()
-                .filter(|target| is_expected_kind(target))
-        });
-        let suggestion = closest_msg(target_name, targets, |t| t.name());
-        anyhow::bail!(
-            "no {} target {} `{}`{}",
+        let targets = packages
+            .iter()
+            .flat_map(|pkg| {
+                pkg.targets()
+                    .iter()
+                    .filter(|target| is_expected_kind(target))
+            })
+            .collect::<Vec<_>>();
+        let suggestion = closest_msg(target_name, targets.iter(), |t| t.name());
+        if !suggestion.is_empty() {
+            anyhow::bail!(
+                "no {} target {} `{}`{}",
+                target_desc,
+                if is_glob { "matches pattern" } else { "named" },
+                target_name,
+                suggestion
+            );
+        }
+        let mut msg = String::new();
+        writeln!(
+            msg,
+            "no {} target {} `{}`.",
             target_desc,
             if is_glob { "matches pattern" } else { "named" },
             target_name,
-            suggestion
-        );
+        )?;
+        if targets.is_empty() {
+            writeln!(msg, "No {} targets available.", target_desc)?;
+            anyhow::bail!(msg);
+        }
+        writeln!(msg, "Available {} targets:", target_desc)?;
+        for target in targets {
+            writeln!(msg, "    {}", target.name())?;
+        }
+        anyhow::bail!(msg);
```

Comment:
nit: these uses early returns / fall throughs when dealing with mutually exclusive high level requirements.  I've found that doing so can make the code harder to read.  Instead highlighting this with if/else's helps draw the readers to it to better understand the structure of the code.

Someone might ask "isn't that banning early returns?".  Not quite.  The difference is in what the code is doing.  For bookkeeping, like handling an error case, you want to draw as little attention to it as possible, so early returns make sense.  When its the core logic of the function, you instead want to call attention to it.  Of course there are different situations where this doesn't always work out in an ideal way but its a starting point I've found useful for more reasdable code.

This applies both to
- `if !suggestions.is_empty()`
- `if targets.is_empty()`

Reply:
Make sense. Addressed.